src/UriInterface.hh

```diff
@@ -271,21 +283,26 @@ interface UriInterface {
   public function withPath(string $path): this;
 
   /**
-   * Return an instance with the specified query string.
+   * Return an instance with the specified query params dict.
    *
    * This method MUST retain the state of the current instance, and return
-   * an instance that contains the specified query string.
+   * an instance that contains the specified query.
    *
-   * Users can provide both encoded and decoded query characters.
+   * Users must provide only unencoded (raw) query characters in order to avoid double
+   * encoding.
    * Implementations ensure the correct encoding as outlined in getQuery().
    *
-   * An empty query string value is equivalent to removing the query string.
+   * All query params which value is set to null MUST be omitted.
+   * The query params dict is assumed to be empty if there are no elements in
+   * it or all params have been omitted.
+   *
+   * An empty query params dict is equivalent to removing the query.
    *
-   * @param string $query The query string to use with the new instance.
-   * @return static A new instance with the specified query string.
-   * @throws \InvalidArgumentException for invalid query strings.
+   * @param dict<string, ?string> $query The query params.
+   * @return static A new instance with the specified query.
+   * @throws \InvalidArgumentException for invalid query params.
    */
-  public function withQuery(string $query): this;
+  public function withRawQuery(dict<string, ?string> $query): this;
```

Comment:
I don't think the `?string` type makes sense: what's the difference between:

- null
- not having the key set
- having it set to the empty string?

Reply:
Ok, seems reasonable. The I guess it worth mentioning that if the value equal to an empty string then the key will be omitted.